design/ceph/object/ceph-cosi-driver.md

```diff
@@ -0,0 +1,161 @@
+
+```yaml
+spec:
+  containers:
+      volumeMounts:
+        - name: cosi-secrets
+          mountPath: /data/cosi
+  volumes:
+  - name: cosi-secrets
+    secret:
+      secretName: ba-secret
+```
+
+```bash
+# cat /data/cosi/bucket_info.json
+```
+
+```json
+{
+      apiVersion: "v1alpha1",
+      kind: "BucketInfo",
+      metadata: {
+          name: "ba-$uuid"
+      },
+      spec: {
+          bucketName: "ba-$uuid",
+          authenticationType: "KEY",
+          endpoint: "https://rook-ceph-my-store:443",
+          accessKeyID: "AKIAIOSFODNN7EXAMPLE",
+          accessSecretKey: "wJalrXUtnFEMI/K...",
+          region: "us-east-1",
+          protocols: [
+            "s3"
+          ]
+      }
+    }
+```
+
+#### Coexistence of COSI and lib-bucket-provisioner
+
+Currently the ceph object store provisioned via Object Bucket Claim (OBC). They both can coexist and can even use same backend bucket from ceph storage. No deployment/configuration changes are required to support both. The lib-bucket-provisioner is deprecated and eventually will be replaced by COSI when it becomes more and more stable. The CRDs used by both are different hence there is no conflicts between them. The user can create OBC and COSI BucketClaim for same backend bucket which have always result in conflicts. Even though credentials for access the buckets are different, both have equal permissions on accessing the bucket. If Rooks creates OBC with `Delete` reclaim policy and same backend bucket is used by COSI BucketClaim with same policy, then bucket will be deleted when either of them is removed.
+
+#### Transition from lib-bucket-provisioner to COSI
+
+This applied to OBC with reclaim policy is `Retain` otherwise the bucket will be deleted when OBC is deleted. So no point in migrating the OBC with `Delete` reclaim policy.
+
+- First the user need to create a **COSI Bucket resource** pointing to the backend bucket.
+- Then user can create BucketAccessClass and BucketAccess using the COSI Bucket CRD.
+- Now the update application's credentials with BucketAccess secret, for OBC it was combination of secret and config map with keys word like AccessKey, SecretKey, Bucket, BucketHost etc. Here details in as JSON format in the [secret](#consuming-cosi-bucket).
+- Finally the user need to delete the existing OBC.
+
+### Points to remember
+
+#### Ceph COSI Driver Requirements
+
+- The credentials/endpoint for the CephObjectStore should be available by creating CephObjectStoreUser with proper permissions
+- The COSI controller should be deployed in the cluster
+- Rook can manage one Ceph COSI driver per Rook operator
+- Rook should not modify COSI resources like Bucket, BucketAccess, BucketAccessClass, or BucketClass.
+
+#### Rook Requirements
+
+- Rook needs to dynamically create/update the secret containing the credentials of the ceph object store for the ceph COSI driver.
```

Comment:
So this for key rotation? This is a current requirement, or a future requirement? If future, please separate the section for current or future requirements.